<launch>
  <arg name="dataset_name" value="EuRoC" />
  <!-- Each dataset has its own set of parameter files -->

  <arg name="data" />

  <arg name="rate" default="1.0" />

  <arg name="parallel" default="False" />

  <!-- Define subscriber topics -->
  <arg name="left_camera_topic" value="/cam0/image_raw"/>
  <arg name="right_camera_topic" value="/cam1/image_raw"/>
  <arg name="imu_topic" value="/imu0"/>
  <!-- Note that we cannot use remap here since offline (parsing rosbag) -->

  <rosparam command="load" file="$(find spark_vio_ros)/param/$(arg dataset_name)/calibration.yaml" />

  <!-- Resiliency Thresholds -->
  <param name="velocity_det_threshold" value="0.1" />
  <param name="position_det_threshold" value="0.3" />
  <param name="stereo_ransac_threshold" value="20" />
  <param name="mono_ransac_threshold" value="30" />

  <!-- Frame IDs for Odometry -->
    <param name="odom_base_frame_id" value="world" />
    <param name="odom_child_frame_id" value="base_link" />

  <!-- More parameter arguments -->
  <arg name="VIO_PARAMS_PATH" value="$(find spark_vio_ros)/param/$(arg dataset_name)/regularVioParameters.yaml" />
  <arg name="TRACKER_PARAMS_PATH" value="$(find spark_vio_ros)/param/$(arg dataset_name)/trackerParameters.yaml" />
  <arg name="VIO_FLAGS_PATH" value="$(find spark_vio_ros)/param/$(arg dataset_name)/flags/StereoVIO.flags" />
  <arg name="MESHER_FLAGS_PATH" value="$(find spark_vio_ros)/param/$(arg dataset_name)/flags/Mesher.flags" />
  <arg name="BACKEND_FLAGS_PATH" value="$(find spark_vio_ros)/param/$(arg dataset_name)/flags/VioBackEnd.flags" />
  <arg name="REGULAR_BACKEND_FLAGS_PATH" value="$(find spark_vio_ros)/param/$(arg dataset_name)/flags/RegularVioBackEnd.flags" />
  <arg name="VISUALIZER_FLAGS" value="$(find spark_vio_ros)/param/$(arg dataset_name)/flags/Visualizer3D.flags" />
  <arg name="VERBOSITY" default="0" />
  <arg name="BACKEND_TYPE" default="1" /> <!-- 1 For regular VIO -->
  <arg name="LOG_OUTPUT" default="False" />
  <arg name="OUTPUT_PATH" default="./" />
  <arg name="VISUALIZE" default="True" /> <!-- Note that this is duplicated from the flags file -->

  <param name="vio_params_filepath" value="$(arg VIO_PARAMS_PATH)" />
  
  <!-- Launch main node  -->
  <node name="stereo_vio_ros_offline" pkg="spark_vio_ros" type="stereo_vio_ros_offline" output="screen" 
        args="--logtostderr=1
              --colorlogtostderr=1 
              --log_prefix=1 
              --vio_params_path=$(arg VIO_PARAMS_PATH) 
              --tracker_params_path=$(arg TRACKER_PARAMS_PATH) 
              --flagfile=$(arg VIO_FLAGS_PATH) 
              --flagfile=$(arg MESHER_FLAGS_PATH)
              --flagfile=$(arg BACKEND_FLAGS_PATH)
              --flagfile=$(arg REGULAR_BACKEND_FLAGS_PATH)
              --flagfile=$(arg VISUALIZER_FLAGS)
              --v=$(arg VERBOSITY)
              --backend_type=$(arg BACKEND_TYPE)
              --log_output=$(arg LOG_OUTPUT)
              --rosbag_path=$(arg data)
              --left_camera_topic=$(arg left_camera_topic)
              --right_camera_topic=$(arg right_camera_topic)
              --imu_topic=$(arg imu_topic)
              --output_path=$(arg OUTPUT_PATH)
              --visualize=$(arg VISUALIZE) 
              --parallel_run=$(arg parallel)"/>

  <!-- launch rviz  -->
  <node type="rviz" name="rviz" pkg="rviz" args="-d $(find spark_vio_ros)/viz/visualize_sparkvio.rviz" />
</launch>
